## **Game Theory**

Two people are flipping fair coins at the same time. What is the probability that person 1 flips 3 heads in a row before person 2 gets 2 heads in a row?


Please simulate this problem to find the solution and give your answer with errors. So the answer should look something like probability +- error.


 * two people
 * coins
 * probability that 1 flips 3 heads in a row, before person 2 gets 2 heads in a row
 
 player01: 1 1 1
   or
 player02: 0 1 1
 
 Probability: 
 Player 1 = player01[1, 1, 1]
 
 
 Player 2 = ( 1 - player02[1, 1])
 
 Tie = player01[1, 1, 1] and player02[0, 1, 1]
 

In [1]:
import numpy as np
from random import randrange

import pandas as pd
from pandas import Series, DataFrame

In [2]:
playerOne = []
playerTwo = []
tie = []


In [3]:
def coinThrow():
    return randrange(2)

In [4]:
def coinGame():
    game = True
    streakOne = 0
    streakTwo = 0
    
    while game:
        coin1 = coinThrow()
        coin2 = coinThrow()
        
        # Check Player 1
        if coin1 == 1:
            streakOne += 1
        else:
            streakOne = 0
        
        # Check Player 2
        if coin2 == 1:
            streakTwo += 1
        else:
            streakTwo = 0
        
        # Check Tie
        if streakOne == 3 and streakTwo == 2:
            # Append Tie
            playerOne.append(0)
            playerTwo.append(0)
            tie.append(1)
            game = False
            streakOne = 0
            streakTwo = 0
            break
        elif streakOne == 3:
            # Player 1 Wins
            playerOne.append(1)
            playerTwo.append(0)
            tie.append(0)
            game = False
            streakOne = 0
            streakTwo = 0
            break
        elif streakTwo == 2:
            # Player 2 Wins
            #dframe2 =  DataFrame({'Player 1': 0, 'Player 2': 1, 'Tie': 0}, index=[0])
            #dframe.append(dframe2, ignore_index = True)
            playerOne.append(0)
            playerTwo.append(1)
            tie.append(0)
            game = False
            streakOne = 0
            streakTwo = 0
            break
        

In [5]:
intLoop = 10000

for i in range(intLoop):
    coinGame()

print(len(playerOne))
print(len(playerTwo))
print(len(tie))

10000
10000
10000


In [6]:
dframe = DataFrame({'Player 1': playerOne, 'Player 2': playerTwo, 'Tie': tie})
dframe

Player 1  Player 2  Tie
0            0         1    0
1            0         0    1
2            0         1    0
3            0         1    0
4            0         1    0
...        ...       ...  ...
9995         0         1    0
9996         0         1    0
9997         0         1    0
9998         0         1    0
9999         1         0    0

[10000 rows x 3 columns]

In [7]:
pro_player1 = sum(playerOne) / intLoop
pro_player2 = sum(playerTwo) / intLoop
pro_tie = sum(tie) / intLoop

print('Probability of player 1 is: ' + str(pro_player1))
print('Probability of player 2 is: ' + str(pro_player2))
print('Probability of player 3 is: ' + str(pro_tie))

Probability of player 1 is: 0.2126
Probability of player 2 is: 0.7376
Probability of player 3 is: 0.0498


### Analysis:
* Anlysis: Based on the collected data, I can conclude the probability of person 1 flips 3 heads in a row before person 2 gets 2 heads in a row is apprxomately 21.3% to 73.8%; where has 0.05% chance of getting tie.